README.md
# Steemvote

This is a program that votes for posts on Steem. It uses piston and thus requires Python 3.

The included script, `steemvoter`, works by monitoring all comments on Steem.
If a comment is eligible for voting, it is stored in a database. The eligible comments
are periodically pulled from the database and voted on after they are a pre-set age.

## Installation

```
$ sudo python3 setup.py install
```

The above command will install all required dependencies.
You can run steemvoter like so: `steemvoter -c path/to/config.json`.

## Configuration

A configuration file specifies the authors to vote on, the key to sign votes with, etc.
Your configuration file can be supplied with a command-line argument (`steemvoter -c path/to/config.json`).
If not supplied, steemvoter will look for the file `steemvote-config.json` or `steemvote-config.yaml` in the current directory.

Example configuration files are located in the [example-configs](./example-configs) directory.

The following config values are required:

- `authors`: A list of authors, and optional values for each (see below).
- `voter_account_name`: The name of your account.
- `vote_key`: The private key used to sign votes with (This should be your "Posting" key).

### Authors

The `authors` list in your config file contains a list of objects.
Each object needs a `name` value, which is the name of the author.

Optional values for each author:

- `priority`: The priority (explained below) that the author has (Default: `normal`).
- `vote_replies`: Whether to vote on all replies the author makes, in addition to their regular posts (Default: `false`).
- `weight`: The weight to vote with. `100.0` represents an upvote, and `-100.0` represents a downvote (Default: `100.0`).

*NOTE* that if you just want to add an author and use the default values, a string can be used instead of an object.

### Delegates

Delegates are other Steem accounts. Steemvoter will watch for votes that your delegates cast,
and will vote for the comments that they vote on.

The `delegates` config list is similar to the `authors` list. However, delegates do not have a `vote_replies` value.

Optional values for each delegate:

- `priority`: The priority (explained below) that comments this delegate votes for have (Default: `normal`).
- `weight`: The weight to vote with. `100.0` represents an upvote, and `-100.0` represents a downvote (Default: `100.0`).

Also like authors, you can use a string instead of an object and the default optional values will be filled in for you.

### Voting Rules

These voting rules are optional, and specify how to vote.

The value `blacklist_categories` can be specified with a list of categories. Comments in those categories will
never be voted for (Default: `[spam]`). Likewise, the value `blacklist_authors` can be specified with a list of author names.
Posts by those authors will never be voted for, even if delegates vote for them.

The following values can be specified as seconds or as human-readable strings (e.g. `"max_post_age": "2 days"`):

- `max_post_age`: The maximum age a post can be for it to be voted on (Default: `2 days`).
- `min_post_age`: The age a post must be before it is voted on (Default: `1 minute`).
- `vote_interval`: The timespan that steemvoter waits between collecting and voting on eligible comments (Default: `10 seconds`).

The following values can be specified as decimals or as human-readable percentages (e.g. `"priority_high": "75%"`):

- `priority_high`: The minimum fraction of voting power that you must have to vote for high priority comments (Default: `80%`).
- `priority_normal`: The minimum fraction of voting power that you must have to vote for normal priority comments (Default: `90%`).
- `priority_low`: The minimum fraction of voting power that you must have to vote for low priority comments (Default: `95%`).

### Database

Steemvoter uses a sqlite database to store data.

By default, the database will be a file called `database.db` in the current directory.
This behavior can be changed using the `database_path` config value.

### RPC

If no RPC options are specified, a public node will be used.

Available RPC options:

- `rpc_node`: The URL of the node to connect to.
- `rpc_user`: RPC username to use.
- `rpc_pass`: RPC password to use.

## Example Configurations

Upvote every post by [@klye](http://steemit.com/@klye), including his replies to other posts.
Wait until posts are 30 seconds old to upvote them.

```
{
    "authors": [
        {"name": "klye", "vote_replies": true}
    ],
    "vote_key": "5XX...",
    "voter_account_name": "myname",
    "min_post_age": 30
}
```

Upvote the posts by [@klye](http://steemit.com/@klye), but not his replies to other posts.
Do not vote if his post is more than 2 hours old.

```
{
    "authors": [
        "klye"
    ],
    "max_post_age": "2 hours",
    "vote_key": "5XX...",
    "voter_account_name": "myname"
}
```


changelog.md
# Changelog

## v0.3.0

* There is now a graphical interface. The GUI is used unless
    the option `-t` or `--terminal` is passed to steemvoter.
    Steemvoter can still be used without installing PyQt4 by passing this option.
* The database now uses sqlite intead of leveldb.
    Leveldb databases are no longer valid.
* Delegates can now be specified with the `delegates` config key.
    Steemvoter will track and vote for comments that your delegates vote for
    as if they were written by your specified authors.
* The config lists `blacklist_authors` and `blacklist_categories` can be used
    to specify authors and categories that should never be voted for.
* Backup authors have been refactored into a per-author priority system.
    Steemvoter will convert your list of backup authors into low-priority
    authors and add them to your main authors list.
* The config keys `priority_high`, `priority_normal`, and `priority_low` are
    used to specify the remaining voting power you must have to vote for
    comments of each priority.
* Steemvoter will replace the config key `max_voting_power` with `priority_low`.
* Steemvoter will replace the config key `min_voting_power` with `priority_high`.

## v0.2.0

* Voting power is tracked. The config key `"min_voting_power"` is used to specify
    the minimum remaining voting power that your account must have for
    steemvoter to vote (Default: `90%`). Current voting power is updated via RPC every 20 seconds.
* A new config key, `"backup_authors"`, is used to specify authors that
    should be voted for when steemvoter is not using enough voting power.
* A new config key, `"max_voting_power"`, is used to specify the remaining voting power
    above which steemvoter will vote for backup authors.
* A new config key, `"vote_interval"`, is used to specify the timespan that
    steemvoter will wait between collecting and voting on eligible comments (Default: `10 seconds`).
* The config key `"vote_delay"` is now `"min_post_age"`. The effect is the same.
    Configurations that still use `"vote_delay"` are backwards-compatible.
* Configuration info is output when steemvoter starts up.
* In the `"authors"` and `"backup_authors"` config lists, a string representing an author name can be used
    instead of an object. In this case, the default values will be used for optional attributes.


steemvote/config.py
import json
import logging
import os

import yaml
import humanfriendly

from steemvote.models import Author, Delegate, Priority

default_values = (
    # Default minimum age of posts to vote on.
    ('min_post_age', 60), # 1 minute.
    # Default maximum age of posts to vote on.
    ('max_post_age', 2 * 24 * 60 * 60), # 2 days.

    # Default high priority voting power.
    ('priority_high', 0.8), # 80%
    # Default normal priority voting power.
    ('priority_normal', 0.9), # 90%
    # Default low priority voting power.
    ('priority_low', 0.95), # 95%

    # Default authors to ignore.
    ('blacklist_authors', []),
    # Default categories to ignore.
    ('blacklist_categories', ['spam']),
)

def get_decimal(data):
    """Parse data into a decimal."""
    if isinstance(data, float):
        return data
    elif data.endswith('%'):
        return float(data.strip('%')) / 100
    # Try to parse a float string (e.g. "0.5").
    if '.' in data:
        return float(data)
    raise ValueError('A percentage or decimal fraction is required')

class ConfigError(Exception):
    """Exception raised when configuration is invalid."""
    pass

class Config(object):
    def __init__(self, no_saving=False):
        # True if unit tests are being run.
        self.no_saving = no_saving
        self.logger = logging.getLogger(__name__)
        self.filepath = ''
        self.config_format = 'json'
        self.options = {}
        self.defaults = {k: v for (k, v) in default_values}

    def get(self, key, value=None):
        """Get a value.

        If no default is specified, the default in self.defaults will
        be used if no value is found.
        """
        result = self.options.get(key, value)
        if result is None and value is None:
            return self.defaults.get(key)
        return result

    def get_decimal(self, key, value=None):
        """Get a value that represents a percentage."""
        val = self.get(key, value)
        try:
            return get_decimal(val)
        except Exception as e:
            raise ConfigError('Invalid config value "%s" for key "%s" (Error: %s)' % (val, key, str(e)))

    def get_seconds(self, key, value=None):
        """Get a value that represents a number of seconds."""
        val = self.get(key, value)
        if isinstance(val, str):
            val = int(humanfriendly.parse_timespan(val))
        return val

    def set(self, key, value):
        self.options[key] = value

    def require(self, key):
        """Raise if a key is not present."""
        if not self.get(key):
            raise ConfigError('Configuration value for "%s" is required' % key)

    def require_class(self, key, cls):
        """Raise if the value of key is not an instance of cls."""
        value = self.get(key)
        if not isinstance(value, cls):
            raise ConfigError('Configuration value for "%s" must be a %s, not %s' % (key, cls.__name__, type(value).__name__))

    def update_old_keys(self):
        """Update old keys for backwards compatibility."""
        updated = False
        # Change "vote_delay" to "min_post_age".
        if self.get('vote_delay') is not None and self.get('min_post_age', -1) == -1:
            self.set('min_post_age', self.get('vote_delay'))
            self.logger.info('Updated old value "vote_delay" to "min_post_age"')
            del self.options['vote_delay']
            updated = True

        # 0.2 used "backup authors" - Convert those to low-priority authors.
        backup_authors = [Author.from_config(i) for i in self.get('backup_authors', [])]
        if backup_authors:
            for author in backup_authors:
                # Add the backup author as a low priority author
                # if it isn't in the main authors list.
                if not self.get_author(author.name):
                    author.priority = Priority.low
                    self.authors.append(author)
            self.logger.info('Updated old value "backup_authors" to low-priority authors')
            del self.options['backup_authors']
            updated = True
        # Change "min_voting_power" to "priority_high".
        if self.get('min_voting_power') is not None and self.get('priority_high', -1) == -1:
            self.set('priority_high', self.get('min_voting_power'))
            self.logger.info('Updated old value "min_voting_power" to "priority_high"')
            del self.options['min_voting_power']
            updated = True
        # Change "max_voting_power" to "priority_low".
        if self.get('max_voting_power') is not None and self.get('priority_low', -1) == -1:
            self.set('priority_low', self.get('max_voting_power'))
            self.logger.info('Updated old value "max_voting_power" to "priority_low"')
            del self.options['max_voting_power']
            updated = True

        if updated:
            self.save()

    def save(self):
        # Return if unit tests are being run.
        if self.no_saving:
            return
        options = dict(self.options)
        options['authors'] = [i.to_dict() for i in self.authors]
        options['delegates'] = [i.to_dict() for i in self.delegates]

        if self.config_format == 'json':
            s = json.dumps(options, indent=4, sort_keys=True)
        elif self.config_format == 'yaml':
            s = yaml.dump(options, indent=4)
        with open(self.filepath, 'w') as f:
            f.write(s)

    def _load_json(self, filepath):
        """Load JSON config."""
        options = {}
        try:
            with open(filepath) as f:
                options = json.load(f)
            self.config_format = 'json'
        except Exception:
            pass
        return options

    def _load_yaml(self, filepath):
        """Load YAML config."""
        options = {}
        try:
            with open(filepath) as f:
                options = yaml.load(f)
            self.config_format = 'yaml'
        except Exception:
            pass
        return options

    def load(self, filepath=''):
        if not filepath:
            filepath = 'steemvote-config.json'
        if not os.path.exists(filepath):
            filepath = 'steemvote-config.yaml'
            if not os.path.exists(filepath):
                return

        if filepath.endswith('.json'):
            options = self._load_json(filepath)
        elif filepath.endswith('.yaml'):
            options = self._load_yaml(filepath)
        self.options = options
        self.filepath = filepath
        self.options_loaded()

    def options_loaded(self):
        self.load_authors()
        self.load_delegates()
        self.update_old_keys()

    def load_authors(self):
        """Load authors from config."""
        authors = self.get('authors', [])
        self.authors = [Author.from_config(i) for i in authors]

    def load_delegates(self):
        """Load delegates from config."""
        delegates = self.get('delegates', [])
        self.delegates = [Delegate.from_config(i) for i in delegates]

    def get_author(self, name):
        """Get an author by name."""
        for author in self.authors:
            if author.name == name:
                return author

    def get_delegate(self, name):
        """Get a delegate by name."""
        for delegate in self.delegates:
            if delegate.name == name:
                return delegate

    def set_authors(self, authors):
        """Set authors and save."""
        if not all(isinstance(i, Author) for i in authors):
            raise TypeError('A list of authors is required')
        self.authors = authors
        self.save()

    def set_delegates(self, delegates):
        """Set delegates and save."""
        if not all(isinstance(i, Delegate) for i in delegates):
            raise TypeError('A list of delegates is required')
        self.delegates = delegates
        self.save()
